clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5858,16 +5858,73 @@ static void handleIntelFPGASimpleDualPortAttr(Sema &S, Decl *D,
                  IntelFPGASimpleDualPortAttr(S.Context, AL));
 }
 
-static void handleIntelFPGAMaxReplicatesAttr(Sema &S, Decl *D,
-                                             const ParsedAttr &A) {
-  checkForDuplicateAttribute<IntelFPGAMaxReplicatesAttr>(S, D, A);
+void Sema::AddIntelFPGAMaxReplicatesAttr(Decl *D, const AttributeCommonInfo &CI,
+                                         Expr *E) {
+  if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
+    // This attribute requires a strictly positive value.
+    if (ArgVal <= 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return;
+    }
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<IntelFPGAMaxReplicatesAttr>()) {
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue());
+      if (DeclExpr && ArgVal != DeclExpr->getResultAsAPSInt()) {
+        Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+        Diag(DeclAttr->getLocation(), diag::note_previous_attribute);
+        return;
+      }
+    }
+    // [[intel::fpga_register]] and [[intel::max_replicates()]]
+    // attributes are incompatible.
+    if (const auto *DeclAttr = D->getAttr<IntelFPGARegisterAttr>()) {
+      Diag(CI.getLoc(), diag::err_attributes_are_not_compatible)
+          << CI << DeclAttr;
+      Diag(DeclAttr->getLocation(), diag::note_conflicting_attribute);
+      return;
+    }
+  }
 
-  if (checkAttrMutualExclusion<IntelFPGARegisterAttr>(S, D, A))
-    return;
+  D->addAttr(::new (Context) IntelFPGAMaxReplicatesAttr(Context, CI, E));
+}
 
+IntelFPGAMaxReplicatesAttr *
+Sema::MergeIntelFPGAMaxReplicatesAttr(Decl *D,
+                                      const IntelFPGAMaxReplicatesAttr &A) {
+  // Check to see if there's a duplicate attribute with different values
+  // already applied to the declaration.
+  if (const auto *DeclAttr = D->getAttr<IntelFPGAMaxReplicatesAttr>()) {
```

Comment:
I think we need some additional test coverage here:
```
// Merging of different arg values.
[[intel::max_replicates(12)]] extern const int var_max_replicates;
[[intel::max_replicates(14)]] const int var_max_replicates = 0; // diagnose

// Merging of incompatible args
[[intel::max_replicates(12)]] extern const int var_max_replicates_2;
[[intel::fpga_register]] const int var_max_replicates_2 = 0;
```
I think the mutual exclusion check is missing from this function.

Reply:
// Merging of incompatible args
[[intel::max_replicates(12)]] extern const int var_max_replicates_2;
[[intel::fpga_register]] const int var_max_replicates_2 = 0;

This is still not working with the current patch. I am working on it. Test case compiles fine without any diagnostic.

Reply:
> Test cases will give errors below:
>
> Non-extern declaration of 'var_max_replicates' follows extern declaration
> 'max_replicates' attribute only applies to local non-const variables and non-static data members

That is a surprise to me: http://cmplrexplorer.intel.com/z/Kr99ez

Looking at the definition of `IntelFPGAConstVar`, it looks like it allows a global constant variable, and this is one of the subjects listed for `IntelFPGAMaxReplicates`. Can you send a Compiler Explorer link of what you're trying and is failing for you?

Reply:
I talked to @AaronBallman offline. This needs a separate lit test since we are testing them as global const var. Also merge function needs to handle this test cases.

Reply:
Done